### ISSUE TITLE
parse_query_string only returns only one instance of a filter specified multiple times.

### ISSUE THREAD

**Issue by dopry** [wagtail/wagtail#10214]
### Issue Summary

Given a query string "term1 term2 filterA:option1 filterA:option2" 
   When pass to wagtail.search.utils.parse_query_string
      Then only filterA:option2 is returned

### Steps to Reproduce

```
from wagtail.search.utils import parse_query_string
parse_query_string( "term1 term2 filterA:option1 filterA:option2")
```
yields
```
({'filterA': 'option2'}, <PlainText 'term1 term2' operator='and' boost=1>)
```

I expect 
```
({'filterA': ['option1', 'option2']}, <PlainText 'term1 term2' operator='and' boost=1>)
```

### Technical details

wagtail: all versions

**Comment by dopry** [comment]
I suspect fixing this may cause some backwards compatibility issues, being that query_string is a term of art with another meaning on the web, maybe a new  function parse_search_string that returns a new structure would be a possible solution. It would be nice if there were ways to specify operators as well for power users.  

Maybe it's worthwhile to consider an implementation based on an existing search syntax like https://github.com/sebastiandev/plyse

**Comment by AbhishekCS3459** [comment]
@dopry  Can I work with this Issue, You Can Assign It to me

**Comment by dopry** [comment]
@AbhishekCS3459 its an open source project you're welcome to submit a PR. Normally issues aren't 'assigned' to people as much as people just take them on and resolve them on their own.  You might wait for a maintainer to weigh in on what approach they think would be good to resolve the issue.

**Comment by AbhishekCS3459** [comment]
@dopry what can be the right suggestion for me, should I wait for the maintainer approach??

**Comment by floels** [comment]
Hi @dopry,

Indeed, given the implementation of the `separate_filters_from_query` function:

https://github.com/wagtail/wagtail/blob/ab05be3bb2d75f7ef2dab227a9b289cd0663a297/wagtail/search/utils.py#L84-L94

only the last value mentioned for a given filter (`filterA: option2` in your example) will be taken into account.

When you say that you would expect the output to be:

`({'filterA': ['option1', 'option2']}, <PlainText 'term1 term2' operator='and' boost=1>)`

this would mean that you would search pages which `filterA` being equal to `'option1'` OR `'option2'`, correct? Can you give us more details about your use case for this?

Cheers,
Florian

**Comment by floels** [comment]
@dopry OK so you want to allow your site's users to search in either body, title, teaser, or a combination of those three?

In that case, perhaps you can form the request to your search endpoint differently, for instance:

`GET /search/?query=word+"my+phrase"&fields=body&fields=title`

and then, in the view that receives this request, do something like:

```python
query = request.GET['query']
fields = request.GET.getlist('fields')
pages_found = MyPage.objects.search(query, fields=fields)
```

leveraging the [`search` function's ability to search only in certain fields](https://docs.wagtail.org/en/stable/topics/search/searching.html#specifying-the-fields-to-search)?

**Comment by gasman** [comment]
It sounds like we could usefully borrow Django's [QueryDict](https://docs.djangoproject.com/en/4.1/ref/request-response/#querydict-objects) class here, so that any existing code that expects filters to be returned from `parse_query_string` as a dict of single values can continue to do so, while any new code that wants to handle multiple values for the same filter can use `.getlist`.

I do also like the idea of adopting something like Plyse so that we aren't reinventing the wheel though!

**Comment by dopry** [comment]
@gasman  would you get behind a solution that implements QueryDict in the current parse_query_string as a fix to this issue. 

While I was excited about plyse last night, it would probably be better to build a broader plan around a plyse implementation and filters/search in general, so that suggestion should maybe be kicked over to an RFC.

**Comment by gasman** [comment]
@dopry Yep, I think a QueryDict solution is the way forward as an immediate fix for this. My interest in Plyse is more prompted by #9951 / #9966 which feels like it's heading towards the need for a real parser - but there's no need to let this issue become blocked by that.

**Comment by dopry** [comment]
@AbhishekCS3459 if you want to address this issue create a feature branch, and update the code useing a QueryDict to store the filters, and update the tests to confirm it supports single and multiple filters, and open a PR.  If you need any guidance on development open a draft PR with your work in progress and ping me. I'll try my best to help you, as I have availability.

**Comment by salty-ivy** [comment]
@AbhishekCS3459  are you still working on it? let me know if you need any help or you wanna transfer this to me?

**Comment by AbhishekCS3459** [comment]
@salty-ivy Yes I currently working in this and will be updating soon.

**Comment by salty-ivy** [comment]
additional context https://github.com/wagtail/wagtail/pull/9966

**Comment by salty-ivy** [comment]
in search/utils.py
 https://github.com/wagtail/wagtail/blob/7a7bbfa6cf11419c9e98569f31dcfa018f0bb7b7/wagtail/search/utils.py#L90
 
you just need to add some conditions

```python
  for match_object in re.finditer(filters_regexp, query_string):
        key, value = match_object.groups()
        if filters.get(key) is None:
            filters[key] = (
                value.strip('"') if value.strip('"') is not value else value.strip("'")
            )

        elif not isinstance(filters.get(key),list):
            filters[key] = [filters[key]]
            filters[key].append(
                value.strip('"') if value.strip('"') is not value else value.strip("'")
            )
        else:
            filters[key].append(
                value.strip('"') if value.strip('"') is not value else value.strip("'")
            )
```
this is working fine locally, make sure to test a bit more rigorously and ofc writing tests for them.
also have another look if you can optimize/ minimise the code logically.

**Comment by gasman** [comment]
Fixed in #10342
